src/Java.Interop/Java.Interop/JniRuntime.JniTypeManager.cs

```diff
@@ -245,13 +245,33 @@ static Assembly TryLoadAssembly (AssemblyName name)
 
 			static bool TryRegisterNativeMembers (JniType nativeClass, Type marshalType, string methods)
 			{
-				var registerMethod  = marshalType.GetTypeInfo ().GetDeclaredMethod ("__RegisterNativeMembers");
+				MethodInfo registerMethod = null;
+				foreach (var m in marshalType.GetRuntimeMethods ())
+					if (m.GetCustomAttribute (typeof (RegisterNativeMembersAttribute)) != null)
+						registerMethod = m;
```

Comment:
Fundamental semantic method here: Do we -- *should we* -- limit to only a single method? That's what this change does: it means a type can only have a single `[RegisterNativeMembers]` method.

Is that *useful*? Is it *desirable*?

It may be useful and desirable, but it's also *not supportable*. The intention of the `[RegisterNativeMembers]` method is to call `JNIEnv::RegisterNativeMethods()` function. `JNIEnv::RegisterNativeMethods()` can only be called *once*. (Not *strictly* true -- you can of course call more than once, but the invocations are *not* "additive"; they're *replacements*. The last invocation "wins.")

As such, allowing multiple `[RegisterNativeMembers]` methods doesn't make sense. It suggests behavior which can't be provided, not unless we change the semantics entirely.

---

Random musings for other purposes, which allowed me to answer the above question.

*Additionally*, there's a performance question. The need for `[RegisterNativeMembers]` is to help [improve xamarin-android process startup time](https://github.com/xamarin/xamarin-android/projects/1). The eventual process would look like:

**Packaging time**:

1. Run `jnimarshalmethod-gen.exe` on *all* assemblies in the app, producing `*-JniMarshalMethod.dll` assemblies. For example, if the app has a `Scratch.MyApp.dll` assembly, packaging-time would produce a `Scratch.MyApp-JniMarshalMethod.dll` assembly. These new assemblies will be included with the app (fast deployment or inclusion in the `.apk`).
2. The `.apk` would contain "normal" Java Callable Wrappers, as currently produced.

**Runtime**:

1. "Eventually," the static constructor for a Java Callable Wrapper will execute, calling `mono.android.Runtime.register()`.
2. `Runtime.register()` is `JNIEnv.RegisterJniNatives()`
3. `JNIEnv.RegisterJniNatives()` will "call" `JniRuntime.CurrentRuntime.TypeManager.RegisterNativeMembers()` -- probably by by having `AndroidTypeManager` override `AndroidTypeManager.RegisterNativeMembers()`.
4. `AndroidTypeManager.RegisterNativeMembers()` will need to:
    1. Try to load the `*-JniMarshalMethod.dll` assembly. if it exists, invoke the `[RegisterNativeMethods]` type located there.
    2. If (4.1) fails, then "fallback" to the current `string.Split()`+ Reflection heavy `JNIEnv.RegisterJniMethods()` implementation.

The *hope* is that using `jnimarshalmethod-gen.exe` and `*-JniMarshalMethod.dll` assemblies will be *faster* than what we currently do, by avoiding the need for `System.Reflection.Emit`.

The question: *is it*?

Unfortunately we can't answer that question without trying.

Meanwhile, it would also be useful to try to imagine *other* ways that JNI method registration could be performed. Is going through `Type.GetType()` + `TypeInfo.GetRuntimeMethods()` + ... going to be efficient? Or do we need some other mechanism for mapping Java types to managed types?

Maybe what we instead/also need to do is use an "int" mapping:

```java
// Java Callable Wrapper
class MyActivity {
    static final int JavaInteropIndex = 0;
    static {
        Runtime.register (JavaInterpIndex, MyActivity.class);
    }
}
```

```csharp
// helper glue in `Mono.Android.dll`?
struct RegistrationInfo {
    Type RuntimeType;
    Action<JniType, string> RegistrationMethod;
    public RegistrationInfo (Type type, Action<JniType, string> method)
    {
        RuntimeType = type;
        RegistrationMethod = method;
    }
}

// ...generated during some packaging step
static class Registrations {
    public static RegistrationInfo[] Info = new[]{
        /* index 0 */ new RegistrationInfo (typeof (MyActivity), MyActivity.RegisterNativeMethods), // method from jnimarshalmethod-gen
        // ...
    };
}
```

This would allow some variation on `JNIEnv.RegisterJniNatives()` to directly turn the `int` index value from `Runtime.register()` into an index into `Registrations.Info`.

This *sounds* cool.

What's unknown is what the startup runtime/overhead penalty will be for the `Registrations` static constructor, which will have *thousands* of `typeof()` expressions + method lookups (off).

Turning this back to the original question -- should we support more than one `[RegisterNativeMethod]` method? -- supporting more than one method would in turn require that `RegistrationInfo` have a `Action<JniType, string>[]`, not a single. Is *that* meaningful? ¯\_(ツ)_/¯

However, supporting more than one method does mean that one could plausibly (but sanely?!) use "custom" method registration alongside `jnimarshalmethod-gen`-based registration.

...which is exactly the problem, and why this *cannot work*.